### PR TITLE
Fix return servicePath in the entity response

### PR DESCRIPTION
**Pull request by chetan-NEC**
**Fix** #2877

- Fix return servicePath in the entity response

### CODE REVIEW

**Comment by chetan-NEC** [comment]
@fgalan, Please review the PR and let me know if you have any suggestions.

**Comment by fgalan** [comment]
Thank you for the PR.

However, I think this is already covered with existing functionality. Let me elaborate:

* In the case of asynchronous context consumption, using notifications, [NGSI payload patching](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#ngsi-payload-patching) can be used to include an attribute (with whichever name the user wants) with the service path information. For instance (from https://github.com/telefonicaid/fiware-orion/blob/master/test/functionalTest/cases/4085_custom_notifications_ngsi_payload/custom_notification_http_ngsi_special_macros.test):

```
        "originalService": {
          "value": "${service}",
          "type": "Text"
        },
```

* In the case of synchronous context consumption (e.g. `GET /v2/entities`) the user specifies the service path in the request (in particular in the `fiware-servicepath` header). Thus, including it in the response is redundant.

What do you think?

**Comment by chetan-NEC** [comment]
I agree with you. However, I am not aware of that because I am new to orion activity. I'll take the necessary precautions to avoid making the same mistakes in the future.

**Comment by fgalan** [comment]
Don't worry, it's not your fault :) It's ours, as we have should have detected this and close parent issue before you start working it...

Thus, do you agree in closing this PR?

**Comment by chetan-NEC** [comment]
Yes, I agree.

**Comment by fgalan** [comment]
Thank you for your understanding!

Issue https://github.com/telefonicaid/fiware-orion/issues/2877 has been also closed.

**Comment by fgalan** [comment]
Taking into account the feedback from @kzangeli at https://github.com/telefonicaid/fiware-orion/issues/2877#issuecomment-1553010495, we are reopening this PR.

@chetan-NEC please do .test based on the case @kzangeli is proposing:

* Usage of multiple elements in fiware-servicepath (e.g. `fiware-servicepath: /A, /B, /C`)
* Usage of wildcards in fiware-servicepath (e.g. `fiware-servicepath: /#`)

Thanks!

**Comment by chetan-NEC** [comment]
As @kzangeli recommended, I am going to look into it.

**Comment by chetan-NEC** [comment]
@fgalan, please provide me some advice on how to use wildcards and multiple elements in the fiware-servicepath.

**Comment by fgalan** [comment]
> @fgalan, please provide me some advice on how to use wildcards and multiple elements in the fiware-servicepath.

The functionally is described here: https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#service-path

Not sure if this answer your question, as it is very broad... In that case, please come back with more specific questions.

**Comment by kzangeli** [comment]
> @fgalan, please provide me some advice on how to use wildcards and multiple elements in the fiware-servicepath.
If I remember correctly, an NGSIv2 entity has only *one* service path, and this issue is about returning it, right?
Why/where would you need to specify more than one?   (that's for queries, not for reponses - if I'm right).

The service path will be (I assume) a field in the json response payload body, per entity, right next to entity id/type

**Comment by chetan-NEC** [comment]
> > @fgalan, please provide me some advice on how to use wildcards and multiple elements in the fiware-servicepath.
> 
> If I remember correctly, an NGSIv2 entity has only _one_ service path, and this issue is about returning it, right? Why/where would you need to specify more than one? (that's for queries, not for reponses - if I'm right).
> 
> The service path will be (I assume) a field in the json response payload body, per entity, right next to entity id/type

Yes, you're correct. The service path should be returned in the entity response, according to this problem. and in this PR, I've added this functionality.

**Comment by fgalan** [comment]
> Yes, you're correct. The service path should be returned in the entity response, according to this issue. and in this PR, I've added this functionality.

You should add more .test to cover the wildcards and list cases. For instance:

01. Create E in SP /Madrid/A
02. Create E in SP /Madrid/B
03. Create E in SP /Barcelona/C
04. Create E in SP /Barcelona/D
05. GET with fiware-service-path: /#
06. GET with fiware-service-path: /Madrid/#
07. GET with fiware-service-path: /Madrid/A,/Barcelona/C
08. GET with fiware-service-path: /Madrid/A,/Barcelona/#
09. GET with fiware-service-path: /Madrid/#,/Barcelona/#

**Comment by fgalan** [comment]
Btw, conflict in CHANGES_NEXT_RELEASE file (trivial to solve) needs to be addressed.

**Comment by chetan-NEC** [comment]
> > Yes, you're correct. The service path should be returned in the entity response, according to this issue. and in this PR, I've added this functionality.
> 
> You should add more .test to cover the wildcards and list cases. For instance:
> 
> 1. Create E in SP /Madrid/A
> 2. Create E in SP /Madrid/B
> 3. Create E in SP /Barcelona/C
> 4. Create E in SP /Barcelona/D
> 5. GET with fiware-service-path: /#
> 6. GET with fiware-service-path: /Madrid/#
> 7. GET with fiware-service-path: /Madrid/A,/Barcelona/C
> 8. GET with fiware-service-path: /Madrid/A,/Barcelona/#
> 9. GET with fiware-service-path: /Madrid/#,/Barcelona/#

Thank you for the suggestion.

**Comment by chetan-NEC** [comment]
@fgalan, In this PR, I've added test cases as you suggested. Please review the PR and let me know if you have any suggestions.

**Comment by fgalan** [comment]
PR looks fine but the conflict in CHANGES_NEXT_RELEASE has to be solved. At the present moment, this is the only line it should appear in that file:

```
- Add: servicePath field to builtin attributes, so return ServicePath in the entity response (#2877)
```

In addition, I'd suggest to add some test testing the new servicePath built-in attribute in notifictions, using `notification.attrs` as `[ "*", "servicePath" ]`.

**Comment by chetan-NEC** [comment]
@fgalan, I have updated this PR with an additional test case, as you suggested.

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1 +1,2 @@
+- Add: servicePath field to builtin attributes, so return ServicePath in the entity response (#2877)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
- Add: servicePath field to builtin attributes (#2877)
```

No need of the last part (in addition, it is uncompleted, as the new builtin attribute is not only for entity responses, but for subscription notifications).

We describe it properly in documentation. I'll do that PR.

File: src/lib/common/globals.h [hunk 2]
```diff
@@ -96,6 +96,7 @@
 #define DATE_EXPIRES    "dateExpires"
 #define ALTERATION_TYPE "alterationType"
 #define ALL_ATTRS       "*"
+#define SERVICE_PATH    "servicePath"
```
Comment (fgalan) [inline src/lib/common/globals.h:99]:
```suggestion
#define SERVICE_PATH    "servicePath"
#define ALL_ATTRS       "*"
```

Better to have the "wildcard" as the last case.

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (fgalan) [reply CHANGES_NEXT_RELEASE:1]: That PR is https://github.com/telefonicaid/fiware-orion/pull/4380

Comment (chetan-NEC) [reply CHANGES_NEXT_RELEASE]: Resolve in c3eec37

File: src/lib/common/globals.h (hunk 2 above)
Comment (chetan-NEC) [reply src/lib/common/globals.h]: Resolve in c3eec37

File: test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+======================================================================
+HTTP/1.1 201 Created
+Date: REGEX(.*)
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Location: /v2/entities/E4?type=T
+Content-Length: 0
+
+
+
+05. GET entities with q parameter and servicePath /Madrid/A
```
Comment (fgalan) [inline test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test:157]: It seems that `q` is not working here... it should return E1 and E3.

I'd suggest to have a look to StringFilter::qMatch and look how other builtins are treated (DATE_MODIFIED, for instance).

File: test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+Date: REGEX(.*)
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Content-Type: application/json
+Content-Length: 2
+
+[]
+
+
+06. GET entities with orderBy servicePath
+=========================================
```
Comment (fgalan) [inline test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test:169]: It seems orderBy is not working. E2 and E4 should bome before E1 and E3.

Maybe it's a matter of fixing sortCriteria() function.

In addition, for completeness plaease include a step for `GET /v2/entities?orderBy=!servicePath`

**Comment by chetan-NEC** [comment]
Hello, @fgalan I would like to request you that let's merge the current PR and open a new issue for this suggested functionality, as q and orderBy arguments are new functionalities for the servicePath built-in attribute. I feel this suggestion implementation would take little more effort.

**Comment by fgalan** [comment]
Builtin attributes should behave as regular attributes and, in this sense, they should properly support `q` and `orderBy` except in justified cases.

Although we could "skip" that support in this PR, that would leave an awful "technical debt" after it (that should be documented, which is also an extra work). Thus, I think it worths the effort to complete the implementation o  `q` and `orderBy` (specially with the hints that I have provided, which should ease the work).

**Comment by chetan-NEC** [comment]
Thank you for your suggestion. I am looking into this thing, as you suggested.

File: test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test (hunk 3 above)
Comment (chetan-NEC) [reply test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test:165]: Fixed in 9e92496.

File: test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test (hunk 4 above)
Comment (chetan-NEC) [reply test/functionalTest/cases/2877_return_ServicePath_in_entity_response/servicePath_behave_with_q_and_orderBy.test:196]: Fixed and added in 9e92496.

**Comment by chetan-NEC** [comment]
@fgalan, please review the PR and let me know if you have any suggestions.

**Review by fgalan — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**